apps/consent-api/src/services/create.ts

```
@@ -0,0 +1,91 @@
+ */
+
+import { ClinicianInviteRequest, ClinicianInviteResponse } from 'types/entities';
+import urlJoin from 'url-join';
+
+import { getAppConfig } from '../config.js';
+
+import axiosClient from './axiosClient.js';
+
+export const createResponse = async ({
+	consentQuestionId,
+	participantId,
+	response,
+}: {
+	consentQuestionId: string;
+	participantId: string;
+	response: boolean;
+}): Promise<any> => {
+	const updateObj = {
+		consentQuestionId,
+		participantId,
+		response,
+	};
+
+	// TODO: POST to data-mapper, then return the result
+	return updateObj;
+};
+
+/**
+ * Creates clinician invite by sending input to data-mapper, which separates data between
+ * Consent and PI DAS and then returns the combined data object from both DB entries
+ * @async
+ * @param data Clinician Invite form input
+ * @returns Created Clinician Invite data as Zod.SafeParseReturnType
+ */
+export const createInvite = async ({
+	participantFirstName,
+	participantLastName,
+	participantEmailAddress,
+	participantPhoneNumber,
+	participantPreferredName,
+	guardianName,
+	guardianPhoneNumber,
+	guardianEmailAddress,
+	guardianRelationship,
+	clinicianFirstName,
+	clinicianLastName,
+	clinicianInstitutionalEmailAddress,
+	clinicianTitleOrRole,
+	consentGroup,
+	consentToBeContacted,
+}: ClinicianInviteRequest): Promise<
+	Zod.SafeParseReturnType<ClinicianInviteResponse, ClinicianInviteResponse>
+> => {
+	const { dataMapperUrl } = getAppConfig();
+	const { data } = await axiosClient.post(urlJoin(dataMapperUrl, 'invites'), {
+		participantFirstName,
+		participantLastName,
+		participantEmailAddress,
+		participantPhoneNumber,
+		participantPreferredName,
+		guardianName,
+		guardianPhoneNumber,
+		guardianEmailAddress,
+		guardianRelationship,
+		clinicianFirstName,
+		clinicianLastName,
+		clinicianInstitutionalEmailAddress,
+		clinicianTitleOrRole,
+		consentGroup,
+		consentToBeContacted,
+	});
+	return ClinicianInviteResponse.safeParse(data);
```

We also need to handle failed requests. Should have a try catch around axios requests to handle unexpected thrown errors, and also have a check for non 200 response codes (axios responses have a `.ok`).

It is good that we are checking for invalid response structures and I really like that we are returning a type that includes a success/failure boolean, but I don't think we need to return the validation info (we just log it so that a developer can fix the problem). When another system is returning invalid responses our service should return an error 500 with a generic "Unexpected server error" message.

There is a really good pattern for functions like this, where on success you return an object of type `{success:true, data: ClinicianInviteResponse}` and on failure you can return `{success:false, message: string}`. We could create a generic type for this response so we don't need to return the Zod parse error, here's a gist I have of a `Result<T>` type that accomplishes this: https://gist.github.com/joneubank/ac5a5b30dafb84915d633131c54aa028

---

Thanks for the gist! I don't see an `ok` field, there is a `statusText` field https://axios-http.com/docs/res_schema which returns the HTTP message for the status code (e.g. `200: "OK"`, `201: "Created"`) so in this case we'd be looking for `statusText === "Created"`. Maybe it's better to just validate the status code is a 201 as we explicitly set in the `data-mapper` endpoint?

The other thing I'm wondering about the `Result` type is if it includes all the information we want. The bad request example I run into a lot is where the participant email I'm creating an invite with is not unique (which is a constraint on it in the `PI` database), so there already exists an invite with that same email address, and that cannot be validated (at least not yet) until we try to create an invite in PI DAS and get a Prisma error. I think that would classify as a 400 error, so I'm wondering what's the best way to return that information from `createInvite` to the router so that it gives the correct error.